feature: move submodule to cmake

meinie0826 opened the pull request:
To support the feature: move submodule to cmake. There are two submodule in the.gitmodules, gtest and json. 
For the json lib, I obtained the nlohmann/json library from the CMakeLists.txt file in the bmf directory using FetchContent.
For the gtest lib, I notice build_aarch64.sh will copy /usr/local/googletest/ to 3rd_party/gtest.

jonahmvp commented on build.sh:
```diff
@@ -116,10 +116,10 @@ done
+# if [[ $LOCAL_BUILD != "OFF" ]];
+# then
+#   git submodule update --init --recursive
+# fi
```
This part can be deleted if all submodule dependencies are successfully removed!

jonahmvp commented on bmf/CMakeLists.txt:
```diff
@@ -141,6 +141,20 @@ endif()
+FetchContent_GetProperties(json)
+if(NOT json_POPULATED)
+    FetchContent_Populate(json)
+endif()
```
Consider moving this section to `cmake/dependencies.cmake`

jonahmvp reviewed (changes requested):
This is a good feature and will help to unify the dependency management to use CMake as much as possible! 

However there are some extra tasks that need to be done for this feature to be complete: 

1. Need to update other occurrences of "3rd_party" in the codebase. 
- https://github.com/BabitMF/bmf/blob/master/build_osx.sh#L160
- https://github.com/BabitMF/bmf/blob/master/cmake/dependencies.cmake#L32
- and more...
2. Need to check if the whole folder `3rd_party` can be deleted. 
- There may be some issues where this will break the build scripts where they try to copy files into the`3rd_party` folder (e.g. https://github.com/BabitMF/bmf/blob/master/build_aarch64.sh#L8-L9). We may need to keep the directory, albeit empty. 
3. Need to also add the FetchContent alternative for gtest (consider a solution similar to glog solution at https://github.com/BabitMF/bmf/blob/master/cmake/glog.cmake). 
4. Delete `.gitmodules`.

meinie0826 commented:
@jonahmvp Thank you for taking the time out of your busy schedule to review my code and provide feedback！
Based on your suggestion, I deleted the previous commit and resubmitted a new one: https://github.com/BabitMF/bmf/commit/0bd5c7ebca25ab26e22e7e86fa23bc172a0e43e9


1. https://github.com/BabitMF/bmf/blob/master/build_osx.sh#L160
    I think this line of code should not be removed because the header file from the JSON library is needed in output/bmf/include/
2. https://github.com/BabitMF/bmf/blob/master/cmake/dependencies.cmake#L32
    I add some code of FetchContent in front of add_library(nlohmann INTERFACE IMPORTED GLOBAL):
```
diff --git a/cmake/dependencies.cmake b/cmake/dependencies.cmake
index 49b07b6..85d0f6d 100644
--- a/cmake/dependencies.cmake
+++ b/cmake/dependencies.cmake
@@ -27,6 +27,20 @@ endif()
 
 ### json
 if (BMF_LOCAL_DEPENDENCIES)
@@ -27,6 +27,20 @@ endif()
 
 ### json
 if (BMF_LOCAL_DEPENDENCIES)
+    include(FetchContent)
+    set(JSON_DOWNLOAD_DIR "${PROJECT_SOURCE_DIR}/3rd_party/json")
+
+    FetchContent_Declare(
+        json
+        GIT_REPOSITORY https://github.com/nlohmann/json.git 
+        GIT_TAG v3.11.2
+        SOURCE_DIR ${JSON_DOWNLOAD_DIR}
+    )
+
+    FetchContent_GetProperties(json)
+    if(NOT json_POPULATED)
+        FetchContent_Populate(json)
+    endif()
     add_library(nlohmann INTERFACE IMPORTED GLOBAL)
     set_property(TARGET nlohmann PROPERTY INTERFACE_INCLUDE_DIRECTORIES
             ${PROJECT_SOURCE_DIR}/3rd_party/json/include)
```
3. maybe `3rd_party` can not be easily removed because of the 3rd-party folder serves as a storage location for many third-party libraries. And retaining it can make the project structure clearer.
4. I have add some code to fetch gtest lib :
```
diff --git a/cmake/gtest.cmake b/cmake/gtest.cmake
index 4025ea7..52fa194 100644
--- a/cmake/gtest.cmake
+++ b/cmake/gtest.cmake
@@ -25,7 +25,14 @@ else()
                 ${GTEST_MAIN_LIBRARY})
         endif()
     else()
-        message("GTEST not found in system path, using builtin module")
-        add_subdirectory(3rd_party/gtest)
+        message("GTEST not found in system path, fetching via FetchContent to ${PROJECT_SOURCE_DIR}/3rd_party/gtes
t")
+        include(FetchContent)
+        FetchContent_Declare(
+            googletest
+            GIT_REPOSITORY https://github.com/google/googletest.git
+            GIT_TAG v1.15.2
+            SOURCE_DIR ${PROJECT_SOURCE_DIR}/3rd_party/gtest
+        )
+        FetchContent_MakeAvailable(googletest)
     endif()
 endif()
```
5. I have removed `.gitmodules`

jonahmvp commented:
Looks good - I successfully compiled your branch on MacOS, so a successful run on CI will be enough to verify your PR! 

Just a few more changes and then should be good to approve: 
- The git submodules indexes don't seem to be fully deleted, try running `git rm --cached 3rd_party/json` and `git rm --cached 3rd_party/gtest` and then commit the result. I agree with your comment on point 3, so maybe add `3rd_party/.gitkeep` to ensure the directory still exists in the repo. 
- I get this deprecation warning due to your usage of `FetchContent_Populate(json)`. Can we use `FetchContent_MakeAvailable(json)` instead? 

```
CMake Warning (dev) at /opt/homebrew/share/cmake/Modules/FetchContent.cmake:1953 (message):
  Calling FetchContent_Populate(json) is deprecated, call
  FetchContent_MakeAvailable(json) instead.  Policy CMP0169 can be set to OLD
  to allow FetchContent_Populate(json) to be called directly for now, but the
  ability to call it with declared details will be removed completely in a
  future version.
Call Stack (most recent call first):
  cmake/dependencies.cmake:42 (FetchContent_Populate)
  CMakeLists.txt:149 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.
```

meinie0826 commented:
@jonahmvp Thank you very much for your guidance and verification.

Under your advice and guidance, I git rm json and gtest and use FetchContent_makeAvailable (json) instead of FetchContent-Populate.I apologize for my carelessness.

meinie0826 commented on cmake/gtest.cmake:
```diff
@@ -25,7 +25,14 @@ else()
-        message("GTEST not found in system path, using builtin module")
-        add_subdirectory(3rd_party/gtest)
+        message("GTEST not found in system path, fetching via FetchContent to ${PROJECT_SOURCE_DIR}/3rd_party/gtest")
+        include(FetchContent)
```
@jonahmvp  Yes, you're right. Then I remove `include(FetchContent)` in `cmake/gtest.cmake` and `bmf/hmp/cmake/dependencies.cmake`. And I move the `include(FetchContent)` to the top in the `cmake/dependencies.cmake` for `hmp/cmake/dependencies.cmake` to use.

jonahmvp reviewed (approved):
LGTM - Nice work 🥇

jonahmvp commented:
The build and test CI action does not run due to the Ubuntu 20.04 LTS runner being deprecated. 

I have verified this PR on Linux by running the CI pipeline from the internal BMF repo. 

Thus, this MR can be merged.
